src/Adapter/EntityTranslation/EntityTranslator.php

```diff
@@ -0,0 +1,223 @@
+     */
+    protected $translator;
+
+    /**
+     * @var string
+     */
+    protected $tableName;
+
+    /**
+     * @var int
+     */
+    protected $shopId;
+
+    /**
+     * @param Db $db
+     * @param TranslatorInterface $translator
+     * @param DataLangCore $dataLang
+     */
+    public function __construct(
+        Db $db,
+        TranslatorInterface $translator,
+        DataLangCore $dataLang
+    ) {
+        //$this->translatorLanguageLoader = new TranslatorLanguageLoader(true);
+        $this->dataLang = $dataLang;
+        $this->db = $db;
+        $this->translator = $translator;
+        $this->tableName = $this->buildTableNameFromDataLang($dataLang);
+    }
+
+    /**
+     * Translate the entity's data in database using reverse translation technique
+     *
+     * @param int $languageId
+     * @param int $shopId
+     *
+     * @throws LanguageNotFoundException
+     * @throws \PrestaShopDatabaseException
+     * @throws \PrestaShopException
+     */
+    public function translate(int $languageId, int $shopId)
+    {
+        $lang = new Language($languageId);
+        if (empty($lang->id)) {
+            throw new LanguageNotFoundException(new LanguageId($languageId));
+        }
+
+        $tableNameSql = bqSQL($this->tableName);
+
+        $shopFieldExists = $this->shopFieldExists($tableNameSql);
+
+        // get table data
+        $sql = "SELECT * FROM `$tableNameSql` WHERE `id_lang` = $languageId"
+            . ($shopFieldExists ? sprintf(' AND `id_shop` = %d', $this->shopId) : '');
+
+        $tableData = $this->db->executeS($sql, true, false);
+
+        if (!empty($tableData)) {
+            $keys = $this->dataLang->getKeys();
+            $fieldsToUpdate = $this->dataLang->getFieldsToUpdate();
+
+            foreach ($tableData as $data) {
+                $updateWhere = [];
+                $updateField = [];
+
+                // Construct update where
+                foreach ($keys as $key) {
+                    $updateWhere[] = '`' . bqSQL($key) . '` = "' . pSQL($data[$key]) . '"';
+                }
+
+                // Construct update field
+                foreach ($fieldsToUpdate as $fieldName) {
+                    if ('url_rewrite' === $fieldName && Language::$locale_crowdin_lang === $lang->locale) {
+                        continue;
+                    }
+
+                    $translatedField = $this->doTranslate($data, $fieldName);
+
+                    if (!empty($translatedField) && $translatedField != $data[$fieldName]) {
+                        $updateField[] = '`' . bqSQL($fieldName) . '` = "' . pSQL($translatedField) . '"';
+                    }
+                }
+
+                // Update table
+                if (!empty($updateWhere) && !empty($updateField)) {
+                    $updateWhere = implode(' AND ', $updateWhere);
+                    $updateField = implode(', ', $updateField);
+
+                    $sql = "UPDATE `$tableNameSql` SET $updateField
```

Comment:
Prepare the string outside the loop, and use sprintf instead. 
The `$shopFieldExists` shouldn't be executed on each loop.

Reply:
I'm refactoring the $shopFieldExists but leaving the string interpolation. AFAIK `sprintf` won't add anything and replacing named tokens with `%s` will decrease readability.